### ISSUE TITLE
Variables versus attributes...

### ISSUE THREAD

**Issue by jklymak** [OceanGlidersCommunity/OG-format-user-manual#12]
Thanks for the hard work sorting this out.  As an outsider, but also an end user:

- Looking at the [draft](https://docs.google.com/document/d/1Ri0V_xJjVRDaKKin0sBFk_6u1Dpt7Xwa/edit), I'm not clear why most of the strings under "Platform Information", "Deployment Information" etc, are variables instead of attributes.  If you try and merge two data sets with different strings here (using say [xarray](https://xarray.org)) I think you will get size mismatches and you won't succeed.  But more fundamentally I don't think of any of these as "variable" for the duration of a single yo.
- I'm also confused about "Parameter Information" and "Geophysical Variables" .  Can't all the "Parameter Information" be stored as attributes on the "Geophysical variables"?  I'm not following the use-case for both.
- It would be helpful to me to see some mock netcdf files in this format.  Its possible that I am misunderstanding some things because I'm not understanding how these are implemented.

Sorry that is three issues!  Thanks again.

**Comment by castelao** [comment]
Hi @jklymak , thank you for your comments and interest.

- Yes I agree, most of those metadata don't sound like variables, but this is just the NetCDF nomenclature. If we ignore the name 'variables' and 'attributes' those are information with dimensions versus information without dimension that applies for the whole dataset. The advantage of having some of these metadata as variables (with dimensions) is to be able to collect different platforms. For instance, one experiment with a fleet of gliders could be grouped in a single file with trajectory dimension size n. In that case, the variable platform would have trajectory as its dimension, so we could identify which glider measured what. Platform information as a global attribute wouldn't allow that. Did I understand your point?
- I don't think that there is a problem with merging two string variables with xarray. If the NetCDF is written as an array of characters, there would be a mismatch on the dimensions (usually str_len*), but as strings, it shouldn't be a problem. In the case of an array of characters, I believe that xarray transforms it into a string and removes the characters' length dimension.
- I already created a mock NetCDF, got some feedback, and just need to manage some time to redo it. I'll keep you posted.
- One last comment. There are a couple of names and structures that I don't like as well, but there is a price on breaking with other conventions and standards. Sometimes it's a hard decision on what is the best. There is a good chance that we missed something, so please let us know your comments and concerns.

**Comment by jklymak** [comment]
Hi @castelao Thanks for the welcome!

>information with dimensions versus information without dimension that applies for the whole dataset. The advantage of having some of these metadata as variables (with dimensions) is to be able to collect different platforms. 

I agree with this, and I think that was my point.  The proposal is that, for instance, `PLATFORM_TYPE` be a variable, but it has no dimensions.  It is probably my ignorance of NetCDF speaking, but I don't see how I would promote it to a dimension, or give it dimensions of "trajectory".  Maybe an example _would_ be helpful.

**Comment by castelao** [comment]
That's a good point. With the Simple Geometries of CF-Conventions, when there is a special case of a dimension size 1 it can be neglected. For instance, a Trajectory Data with a single trajectory can be written with trajectory as a scalar such as [H.4.2](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.9/cf-conventions.html#_single_trajectory). But I see your point and agree with you. On the OG-1.0 manual, the dimensions should be defined explicitly to avoid any confusion. Something like: `PLATFORM_TYPE(trajectory)`. And following CF, a special case of a single trajectory (which in our case would be actually the most common case),  the PLATFORM_TYPE could be downgraded to a scalar. If that special case is stacked with more trajectories, the PLATFORM_TYPE would be upgraded to a vector with dimension trajectory.

I agree with you. Let's see what the others think about this.

**Comment by gkrahmann** [comment]
Hi Guilherme and Jody and others,

I have been wondering about this variable vs attribute issue since the first versions of the then EGO netcdf format presented in Trieste in 2013. But now that you wrote that it should also be able to handle fleets, I can at least see where the variable approach comes from. Still, I am not a friend of it. To me it is a twisting of the idea behind netcdf files. I would keep things simple with one file per glider deployment which is something like a 'natural unit'.

When I look at the google doc, it is not yet consistent with the fleet idea. 
There is no 'trajectory' dimension.
And Victor 3 weeks ago removed the variable-version of the wmoid
in favor of the attribute. So what is the goal now ?  Fleets or single deployments ? 

Once that has been decided the consequences are clear. For a fleet basically all glider dependent information has to be in variables. 

At least when the information is a string. (Though for numerical values netcdf attributes could also hold multiple values.)
Actually the contributor_* global attributes are an example for combining multiple strings in a single attribute. Why not follow that way for all these strange attribute-variables ?  

The fleet approach also has other consequences not visible in the document yet:
What is e.g. the 'id' when platform_code has multiple strings ?
You would need a 'replacement' for the platform_code in the 'id' definition.

All these problems stem from the idea of having multiple deployments in a single netcdf file. Drop that and everything will be much easier. And stick 'attributes' into attributes and not into variables ;-)

just my 5c (maybe a bit more than 5)
Gerd

**Comment by jklymak** [comment]
I agree with the above.  The root question is what is the expected use of the files.  I could see concatenating timeseries files to make a long time series.  However if I do that, and the platform changes, it is not much use to have "platform" as a singleton variable.  It either needs to have the same dimension of time, or I drop the info   In either case it may as well just be an attribute.  

Similarly if I grid by profile.  In that case I may have profile_time and depth as coordinates, in which case again not much is gained by keeping platform as a variable.

**Comment by justinbuck** [comment]
We met on this on 14th June 2022, @castelao is working on an example that will give us the variables option for comparison at a follow up meeting

**Comment by jklymak** [comment]
Examples of the proposed format would be a lot easier to review than the actual manual.  Looking forward to this!

**Comment by vturpin** [comment]
During our last meeting we discussed the remaining issue that was really blocking the release of OG1.0.

Considering that the focused data set granularity is the glider mission, the description of the data set with metadata should be made easy for this level of granularity.
Considering that data set aggregation of glider mission is not the N°1 priority of this format and when aggregated, there will be technical solutions (not simple though) to store the individual metadata and not losing it.
Even though we are loosing the "100% self descriptive capacity of netcdf format", we have reached a consensus to move to a solution where metadata are stored under variable attribute instead of variables.

Note that all participant agree on this solution : 7 votes against 0.

This decision will impact the format description and the cdl.
